count_mat_ATC2 = readRDS(url('http://pklab.med.harvard.edu/teng/data/count_mat_ATC2.rds'))
df_allele_ATC2 = readRDS(url('http://pklab.med.harvard.edu/teng/data/df_allele_ATC2.rds'))

# run
out = run_numbat(
  count_mat_ATC2, # gene x cell integer UMI count matrix 
  ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
  df_allele_ATC2, # allele dataframe generated by pileup_and_phase script
  genome = "hg38",
  t = 0.1,
  nu = 1, 
  plot = TRUE,
  out_dir = '../../Datasets/CopyKAT/analysis/Numbat/ATC2_1/'
)


nb_ATC2 = Numbat$new(out_dir = '../../Datasets/CopyKAT/analysis/Numbat/ATC2_1/')
mypal = c('1' = 'gray', '2' = "#377EB8", '3' = "#4DAF4A", '4' = "#984EA3", '5' = 'red','6' = 'blue')
nb_ATC2$plot_phylo_heatmap(
  clone_bar = TRUE,
  p_min = 0.9,
  pal_clone = mypal
)
